File: CS/Form1.cs
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Base;
using System;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace dxExample {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
            gridControl1.DataSource = GetData(10);
            gridView1.RowCellStyle += gridView1_RowCellStyle;
            gridView1.CellValueChanged += gridView1_CellValueChanged;
        }

        List<GridCell> modifiedCells = new List<GridCell>();
        bool CellExists(int sourceRowIndex, GridColumn col) {
			GridCell existingCell = modifiedCells.Where(c => c.Column == col && c.RowHandle == sourceRowIndex).FirstOrDefault();
            return existingCell != null;
		}

        void gridView1_CellValueChanged(object sender, DevExpress.XtraGrid.Views.Base.CellValueChangedEventArgs e) {
            if (!CellExists(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column)){
                modifiedCells.Add(new GridCell(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column));
			}
        }

        void gridView1_RowCellStyle(object sender, DevExpress.XtraGrid.Views.Grid.RowCellStyleEventArgs e) {
            if (CellExists(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column)) {
                e.Appearance.BackColor = Color.ForestGreen;
                e.Appearance.BackColor2 = Color.LimeGreen;
                e.Appearance.Font = new Font(e.Appearance.Font, FontStyle.Bold);
			}
        }

        DataTable GetData(int rows) {
            DataTable dt = new DataTable();
            dt.Columns.Add("ID", typeof(int));
            dt.Columns.Add("Info", typeof(string));
            for (int i = 0; i < rows; i++)
                dt.Rows.Add(i, "Info" + i);
            return dt;
        }

        private void clearButton_Click(object sender, EventArgs e) {
            modifiedCells.Clear();
            gridView1.LayoutChanged();
        }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Base
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms

Namespace dxExample
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
			gridControl1.DataSource = GetData(10)
			AddHandler gridView1.RowCellStyle, AddressOf gridView1_RowCellStyle
			AddHandler gridView1.CellValueChanged, AddressOf gridView1_CellValueChanged
		End Sub

		Private modifiedCells As New List(Of GridCell)()
		Private Function CellExists(ByVal sourceRowIndex As Integer, ByVal col As GridColumn) As Boolean
			Dim existingCell As GridCell = modifiedCells.Where(Function(c) c.Column Is col AndAlso c.RowHandle = sourceRowIndex).FirstOrDefault()
			Return existingCell IsNot Nothing
		End Function

		Private Sub gridView1_CellValueChanged(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CellValueChangedEventArgs)
			If (Not CellExists(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column)) Then
				modifiedCells.Add(New GridCell(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column))
			End If
		End Sub

		Private Sub gridView1_RowCellStyle(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.RowCellStyleEventArgs)
			If CellExists(gridView1.GetDataSourceRowIndex(e.RowHandle), e.Column) Then
				e.Appearance.BackColor = Color.ForestGreen
				e.Appearance.BackColor2 = Color.LimeGreen
				e.Appearance.Font = New Font(e.Appearance.Font, FontStyle.Bold)
			End If
		End Sub

		Private Function GetData(ByVal rows As Integer) As DataTable
			Dim dt As New DataTable()
			dt.Columns.Add("ID", GetType(Integer))
			dt.Columns.Add("Info", GetType(String))
			For i As Integer = 0 To rows - 1
				dt.Rows.Add(i, "Info" & i)
			Next i
			Return dt
		End Function

		Private Sub clearButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles clearButton.Click
			modifiedCells.Clear()
			gridView1.LayoutChanged()
		End Sub
	End Class
End Namespace
